proposals/052-cryptography-leg-1.md

```diff
@@ -0,0 +1,329 @@
+
+The first leg of this proposal is for the development of high-quality, production-ready bindings to the Botan C++ cryptography library, and from acceptance to release on Hackage is estimated to take 3 months of full-time development. We see several challenges but no significant technical risk to this proposal; known challenges involve strictness and developing additional bindings to C++; these items are of high-value and may take significant effort to resolve, but are known to be solvable. Considerable progress has already been made, and the proposed funding will help ensure its completion.
+
+# Background
+
+Cryptography is an essential part of modern computing and information security, but it requires care, nuance, and knowledge to properly implement, and is extremely easy to mess up. It covers a diverse range of functions, with individual algorithms often having their own particular quirks that must be acknowledged to properly and safely use. Failure to properly implement cryptography carries the risk of severely compromising the security of any system in which it is used. Manually implementing common cryptographic primitives such as hashes, ciphers, digital signatures, big-integer arithmetic is thus considered extremely unwise, both in terms of security and performance.
+
+Cryptography is even harder for functional programming languages. Properly implementing techniques such as sanitizing memory, performing operations in constant time, are already difficult enough in languages such as C, and in a lazy functional language it is made even more difficult by such things as the potential for references to be kept alive within thunks. As a result, the space of functional cryptography is relatively under-explored and under-developed, having favored other languages where such things are easier to manage.
+
+However, if we accept this downside as a part of a trade-off, and explore the ways that functional programming is beneficial to cryptography (monads, type systems, referential transparency), there is then also the potential for functional cryptography to be made *easier*. The functional machinery of Haskell is well-suited for adding important contextual control to cryptography, and can help prevent or eliminate many issues and errors in ways that languages such as C cannot. I believe that there is room for significant improvement, given a directed effort to develop the space.
+
+# Problem Statement
+
+Cryptography in Haskell lacks significant capability beyond basic primitives. This places a significant burden on developers to properly implement various security techniques, and exposes end users to significant risk in the event of a lapse in security. Some companies have built their own solution to this, but there is no community-driven, community-owned solution.
```

Comment:
There is nothing wrong with a company building it per se\*, rather the issue is of ownership of the resulting ecosystem, and the resulting incentive misalignment:

1) cryptography and security are a preventative measure and a cost center for most companies
2) companies for which cryptography / security are a product have a strong incentive to marshal users into a walled garden at first in the name of security, but then later in the name of profit
3) Developing an ecosystem is a significant up-front investment, and so investors may expect ownership of the ecosystem in order to recoup their cost
4) There is an incentive to wait for someone else to develop an open ecosystem, instead of expending resources yourself to do it

As such, for-profit approaches are incentivized to either invest in a private ecosystem, or invest in an established open ecosystem, but are perversely incentivized against building a new open ecosystem.

For reference, technically *I* am a company here, as I have my own LLC set up in order to protect my intellectual property, due to a previous employer having built a product based on it behind my back while acting interested in investing in my work - it was a harsh lesson.

Not all companies are like that, but the profit incentive is exceedingly strong, and so care must be taken to outline the ownership of the resulting ecosystem, in order to engender trust. I should make some revisions here to better elucidate this.